test_register_filter fails on some (32bit?) arches: OverflowError: Python int too large to convert to C ssize_t

debian builds of h5py 3.9.0 are failing on some less common architectures, see the red blocks at https://buildd.debian.org/status/package.php?p=h5py

Arches affected by the same problem are  armel, i386 (i686),  hppa,  powerpc,  sparc64.  armhf is in progress.
edit: also affected: armhf, x32.

The build log for i386 for instance is https://buildd.debian.org/status/fetch.php?pkg=h5py&arch=i386&ver=3.9.0-1&stamp=1695926469&raw=0

The error occurs in test_register_filter (test_h5z.py):
```
_____________________________ test_register_filter _____________________________

    def test_register_filter():
        filter_id = 256  # Test ID
    
        @H5ZFuncT
        def failing_filter_callback(flags, cd_nelemts, cd_values, nbytes, buf_size, buf):
            return 0
    
        dummy_filter_class = H5ZClass2T(
            version=h5z.CLASS_T_VERS,
            id_=filter_id,
            encoder_present=1,
            decoder_present=1,
            name=b"dummy filter",
            can_apply=None,
            set_local=None,
            filter_=failing_filter_callback,
        )
    
>       h5z.register_filter(addressof(dummy_filter_class))

build/h5py/_debian_h5py_serial/tests/test_h5z.py:65: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
h5py/_debian_h5py_serial/_objects.pyx:54: in h5py._debian_h5py_serial._objects.with_phil.wrapper
    ???
h5py/_debian_h5py_serial/_objects.pyx:55: in h5py._debian_h5py_serial._objects.with_phil.wrapper
    ???
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

>   ???
E   OverflowError: Python int too large to convert to C ssize_t

h5py/_debian_h5py_serial/h5z.pyx:102: OverflowError
=========================== short test summary info ============================
FAILED build/h5py/_debian_h5py_serial/tests/test_h5z.py::test_register_filter
=========== 1 failed, 709 passed, 28 skipped, 3 deselected in 9.62s ============
```
I might have thought it was an issue with 32-bit architectures, except that sparc64 is 64-bit. The build on the 32-bit arch armhf is in progress. [Edit: armhf and x32 also affected]


```
Summary of the h5py configuration (i386)
---------------------------------

h5py    3.9.0
HDF5    1.10.10
Python  3.11.5 (main, Aug 29 2023, 15:31:31) [GCC 13.2.0]
sys.platform    linux
sys.maxsize     2147483647
numpy   1.24.2
cython (built with) 0.29.36
numpy (built against) 1.24.2
HDF5 (built against) 1.10.10
```

--- a/h5py/h5z.pyx
+++ b/h5py/h5z.pyx
@@ -11,6 +11,7 @@
     Filter API and constants.
 """
 
+from libc.stdint cimport uintptr_t
 from ._objects import phil, with_phil
 
 
@@ -99,7 +100,7 @@ def get_filter_info(int filter_code):
 
 
 @with_phil
-def register_filter(Py_ssize_t cls_pointer_address):
+def register_filter(uintptr_t cls_pointer_address):
     '''(INT cls_pointer_address) => BOOL
 
     Register a new filter from the memory address of a buffer containing a

